### ISSUE TITLE
IdSrv (unexpectedly) requesting offline scope from Google?

### ISSUE THREAD

**Issue by RPM1984** [IdentityServer/IdentityServer3#2373]
So i'm using the OWIN MW to login to IdSrv via Google/Facebook.

Here's the scopes for my Client (all of which i'm passing in my signin request):

```
AllowedScopes = new List<string>
{
	Constants.StandardScopes.OpenId,
	Constants.StandardScopes.Email,
	Constants.StandardScopes.Profile,
	Constants.StandardScopes.Roles,
	"myApi"
},
```

Note no offline access.

Now, when i did the Google login, i was first asked to consent to profile/email/etc which is totally expected. After consenting, _another_ consent screen was displayed, asking for Offline access? 

Firstly, weird it didn't include it on the first consent screen.

If i consent, it's all good - i get back to IdSrv, but subsequent logins _keep_ asking just for offline access. 

Here's the URL for the consent screen: (URL decoded, some values removed for brevity)
```
https://accounts.google.com/o/oauth2/auth?scope=openid profile email&response_type=code&redirect_uri=https://localhost:44301/core/signin-google&state=longstatevalue&client_id=xxxxx.apps.googleusercontent.com&hl=en-GB&from_login=1&as=-58cbb5e5d3f08530&authuser=0
```
 
 Note how it's only asking for `openid profile email` scopes, which is as expected. So that URL/scopes looks good to me?

So why is Google asking the user (me) for offline access?

**Comment by mderriey** [comment]
I believe that the scopes you allow for a specific client via the `AllowedScopes` property only defines which scopes the client can ask to IdSrv. The scopes that are requested for external providers are configured by the `AuthenticationOptions.IdentityProviders` property.

For example, the [`GoogleOAuth2AuthenticationOptions`](https://msdn.microsoft.com/en-us/library/microsoft.owin.security.google.googleoauth2authenticationoptions(v=vs.113).aspx) has a `Scope` property which defines which scopes will be requested by the Google application you created.

Both are completely independent. Your client application delegates authentication to idSrv. And if a user chooses an external identity provider, then consider that IdSrv delegates authentication to, let's say, Google.
If the user authenticates successfully against an external identity provider, the user is considered to be authenticated, but he'll still have to accept that your client application can access whatever scopes you defined.

Hope that is clear.
In short, check the settings for the external identity providers.

**Comment by mderriey** [comment]
After looking [at the source code for `GoogleOAuth2AuthenticationOptions`](http://katanaproject.codeplex.com/SourceControl/latest#src/Microsoft.Owin.Security.Google/GoogleOAuth2AuthenticationOptions.cs) and by the comments, it could be worth looking at the `AccessType` property, too. Comment states: `access_type. Set to 'offline' to request a refresh token.`

**Comment by RPM1984** [comment]
@mdierry thanks for your reply. I get what you're saying, but it shouldn't be like that? The whole point is to delegate authentication, all the client cares about is getting back the scopes it requests. If you're saying it's configured elsewhere, what's the point of the client passing the scopes at all?

**Comment by mderriey** [comment]
There are 2 levels of delegation in this case.
Your client application delegates to IdSrv, but IdSrv delegates to Google. That second part is independent from the first, so the scopes involved can be - and most likely are - different.

Let's say the user authenticates against Google, IdSrv receives the claims. It could then check in the local user store for a user which email matches the one Google sent.

Then the usual flow continues, and the user is asked to provide consent for the scopes the client application asked to IdSrv.

Check your Google settings, maybe something isn't right.

**Comment by RPM1984** [comment]
@mderriey 

From [here](https://developers.google.com/identity/protocols/OAuth2WebServer?hl=ja)

> access_type	online or offline	Indicates whether your application needs to access a Google API when the user is not present at the browser. This parameter defaults to online. If your application needs to refresh access tokens when the user is not present at the browser, then use offline. This will result in your application obtaining a refresh token the first time your application exchanges an authorization code for a user.

Nothing in the URL - so it's online, not offline.

I also tried forcing the value in the settings:

```
app.UseGoogleAuthentication(new GoogleOAuth2AuthenticationOptions
{
	AuthenticationType = "Google",
	Caption = "Google",
	SignInAsAuthenticationType = signInAsType,
	ClientId = "secret.apps.googleusercontent.com",
	ClientSecret = "secret",
	AccessType = "online"
});
```

Which then generates the URL:

> https://accounts.google.com/o/oauth2/auth?access_type=online&scope=openid profile email&response_type=code&redirect_uri=https://localhost:44301/core/signin-google&state=xxx&client_id=xxx.apps.googleusercontent.com&hl=en-GB&from_login=1&as=46a7725ebc1e6fca&authuser=0

But it's still prompting for offline access.

Not sure what "Google settings" i can check..the API console is only regarding client ids/secrets and URI's, nothing to do with scopes. So it feels like i need to do something in my OWIN setup, or in IdSrv..

@brockallen @leastprivilege any ideas?

**Comment by RPM1984** [comment]
The weird thing is, i de-authorized the app access and started again.

Worked fine - first consent screen asked for profile email and identity, then i accepted, then i'm back to the client (no offline access/secondary screen).

So that's totally cool, no issues.

But then after logging out then trying again, it asks for offline access only, and always does that from now on...

**Comment by RPM1984** [comment]
Couple of interesting SO questions...

http://stackoverflow.com/questions/21405274/this-app-would-like-to-have-offline-access-when-access-type-online
http://stackoverflow.com/questions/21097008/the-app-keeps-asking-for-permission-to-have-offline-access-why

**Comment by leastprivilege** [comment]
Google scopes are configured on the MS Google middleware - not in idsrv.

Is this resolved now?

**Comment by RPM1984** [comment]
@leastprivilege i thought IdSrv converts the `signInMessage` on the `/external?provider=Google` to the scopes required? I'm not setting any scopes in the MS Google MW, so the original scopes the client requested is being passed onto Google (as expected). 

However, i'm not sure which area the issue is in, so I [raised a Q on SO] (http://stackoverflow.com/questions/34666105/google-oauth2-asking-for-offline-access). Happy for you to close this if you feel it's unrelated to IdSrv.

**Comment by leastprivilege** [comment]
We don't pass anything on. It is purely a matter of configuring the external provider middleware

**Comment by RPM1984** [comment]
@leastprivilege but if i have a MW config like this:

```
          app.UseGoogleAuthentication(new GoogleOAuth2AuthenticationOptions
            {
                AuthenticationType = "Google",
                Caption = "Google",
                SignInAsAuthenticationType = signInAsType,
                ClientId = "xxx.apps.googleusercontent.com",
                ClientSecret = "xxx"
            });
```

(no scope configuration).

Then how does the URL get created with the "openid profile email" scopes? Is that the default?

**Comment by brockallen** [comment]
You need to configure the scopes requested from google -- it's your google MW that's you're configuring. We don't really know anything about it -- we just know the MW's name is "google".

**Comment by RPM1984** [comment]
It was a localhost issue. Once on server, was fine. Weird.

**Comment by roblapp** [comment]
@brockallen In regards to @RPM1984 's question - how does IdentityServer know how to create the correct auth  URL that gets sent to Google? For example, https://accounts.google.com/o/oauth2/auth?response_type=code&client_id=theClientId&redirect_uri=https://app.com/callback&scope=openid profile email&state=theState ... how do the scopes get set there?

**Comment by brockallen** [comment]
it's all from the msft middleware

**Comment by sitefinitysteve** [comment]
@RPM1984 i don't suppose you could share the exact fix?  Users are freaking out over this :)

**Comment by RPM1984** [comment]
@siteinfinitysteve it only occurred on local host and not on any server environments. So if you're getting it there, it must be different issue. Sorry
